base/primes.jl

```diff
@@ -6,31 +6,43 @@
 #     http://thomasinterestingblog.wordpress.com/2011/11/30/generating-primes-with-the-sieve-of-atkin-in-c/
 #     http://dl.dropboxusercontent.com/u/29023244/atkin.cpp
 #
-function primesmask(s::AbstractVector{Bool})
-    n = length(s)
-    n < 2 && return s; s[2] = true
-    n < 3 && return s; s[3] = true
-    r = isqrt(n)
-    for x = 1:r
-        xx = x*x
-        for y = 1:r
-            yy = y*y
-            i, j, k = 4xx+yy, 3xx+yy, 3xx-yy
-            i <= n && (s[i] $= (i%12==1)|(i%12==5))
-            j <= n && (s[j] $= (j%12==7))
-            1 <= k <= n && (s[k] $= (x>y)&(k%12==11))
+function primesmask(sieve::AbstractVector{Bool})
+    limit = length(sieve)
+    limit < 2 && return sieve; sieve[2] = true
+    limit < 3 && return sieve; sieve[3] = true
+    r = isqrt(limit)    # highest factor
+    for x = 1:r; x² = x*x
+        for y = 1:r; y² = y*y
+            n = 4x² + y²
+            if n ≤ limit && (n % 12 == 1 || n % 12 == 5)
+                sieve[n] = !sieve[n]
+            end
+            n = 3x² + y²
+            if n ≤ limit && n % 12 == 7
+                sieve[n] = !sieve[n]
+            end
+            n = 3x² - y²
+            if x > y && n ≤ limit && n % 12 == 11
```

Comment:
I see, I wasn't really paying attention to the `x > y` condition in there (which guaranties n is not negative). So the index will be always in bounds given the conditions. In the case of the previous implementation is was really necessary.